libclamav_rust/src/fmap.rs

```diff
@@ -0,0 +1,102 @@
+/*
+ *  Rust interface for libclamav FMap module
+ *
+ *  Copyright (C) 2023 Cisco Systems, Inc. and/or its affiliates. All rights reserved.
+ *
+ *  Authors: Micah Snyder
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License version 2 as
+ *  published by the Free Software Foundation.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ *  MA 02110-1301, USA.
+ */
+
+use std::convert::TryFrom;
+
+use log::{debug, error};
+use thiserror::Error;
+
+use crate::{sys, util::optional_cstring};
+
+/// FMapError enumerates all possible errors returned by this library.
+#[derive(Error, Debug)]
+pub enum FMapError {
+    #[error("Invalid parameter: {0}")]
+    InvalidParameter(String),
+
+    #[error("{0} parmeter is NULL")]
+    NullParam(&'static str),
+
+    #[error("Offset {0} and length {1} not contained in FMap of size {2}")]
+    NotContained(usize, usize, usize),
+
+    #[error("FMap pointer not initialized: {0}")]
+    UninitializedPtr(&'static str),
+}
+
+#[derive(PartialEq, Eq, Hash, Debug)]
+pub struct FMap {
+    fmap_ptr: *mut sys::cl_fmap_t,
+}
+
+impl TryFrom<*mut sys::cl_fmap_t> for FMap {
+    type Error = &'static str;
```

Comment:
It's generally a good idea to utilize an error type that implements `std::error::Error`, as it guarantees certain behaviors that will be expected upstream.  `&str` kinda works because it implements `Display`, which is one of `std::error::Error`'s bounds, but it doens't implement the entire trait.

Reply:
If I do
```rust
impl TryFrom<*mut sys::cl_fmap_t> for FMap {
    type Error = dyn std::error::Error;
```
then I get
```
error[E0277]: the size for values of type `(dyn std::error::Error + 'static)` cannot be known at compilation time
   --> libclamav_rust/src/fmap.rs:54:48
    |
54  |     fn try_from(value: *mut sys::cl_fmap_t) -> Result<Self, Self::Error> {
    |                                                ^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `Sized` is not implemented for `(dyn std::error::Error + 'static)`
note: required by a bound in `Result`
   --> /home/micah/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:502:20
    |
502 | pub enum Result<T, E> {
    |                    ^ required by this bound in `Result`

```

Reply:
```suggestion
#[derive(Debug, thiserror::Error)]
pub enum FmapError {
    #[error("Attempted to create Rust FMap interface from NULL pointer")]
    Null,
}
```